### ISSUE TITLE
Longpress to activate drag on item

### ISSUE THREAD

**Issue by FemtioKiloKokain** [isaacHagoel/svelte-dnd-action#338]
Hi 👋 

First of, this tool has been super helpful for us and and has been really great to work with. 

I have an orderable list which on mobile uses the entire screen width. To not interfere with scrolling I've had to use a scrollhandle to trigger drag. Those handles do interfere somewhat with the design though, so instead I've been trying to make the area draggable only after a longpress (500ms) on an item but haven't had much success. Is this something that would be possible to achieve?

**Comment by isaacHagoel** [comment]
Probably. see the drag handle example that manipulates "dragDisabled".
If you make a REPL that illustrates your situation i am happy to have a look

**Comment by FemtioKiloKokain** [comment]
Thanks for your reply! 

I've tried to do something similar to the drag handle, and i'm able to manipulate dragDisabled on longpress but it seems to lose grab of the item. 

I've put together this https://replit.com/@FemtioKiloKokai/svelte-dnd-action-longpress#src/App.svelte if thats makes things clearer.

**Comment by isaacHagoel** [comment]
I suddenly recall i had made this before and it might help you
https://svelte.dev/repl/7e28249932d84d59a9d01504bf14417e?version=3.44.3
there some missing bits which i am happy to help with but see if it works for u

**Comment by isaacHagoel** [comment]
@FemtioKiloKokain any updates on this? did the link I shared help?

**Comment by madeleineostoja** [comment]
I know this is an ancient issue but I ran up against it too, would be great to get that demo in the readme for anyone looking for built in long press support on mobile (or even better, add a `dragDelay` or similar option baked in)

**Comment by FemtioKiloKokain** [comment]
Sorry for never replying to this, but I actually got it working. One thing I noticed is that having the delay set to 500ms makes it clash with android context menu, which had me confused for a long time, and I didn't find any way around that other than lowering the delay. Other than that, my implementation ended up as almost a copy of your demo.

So thank you @isaacHagoel, your help was much appreciated. And I agree with @madeleineostoja, while drag handles might be an alternative in many cases, having longpress as a built in feature would be great since it's quite a common use case.

**Comment by isaacHagoel** [comment]
@FemtioKiloKokain if you can make a REPL with your solution I will happily to add to the README

**Comment by PlopTheReal** [comment]
> Sorry for never replying to this, but I actually got it working. One thing I noticed is that having the delay set to 500ms makes it clash with android context menu, which had me confused for a long time, and I didn't find any way around that other than lowering the delay. Other than that, my implementation ended up as almost a copy of your demo.
> 
> So thank you @isaacHagoel, your help was much appreciated. And I agree with @madeleineostoja, while drag handles might be an alternative in many cases, having longpress as a built in feature would be great since it's quite a common use case.

Hey, how did you make it work then? I got the code in place with touchstart, touchmove and touchend but whenever the touchmove kicks in, I can't find a way to 'aborts' the slide of the window.

**Comment by isaacHagoel** [comment]
Normally preventing defailt on touch start does that. If u share a repl it
would be easier to see what your exact problem is

**Comment by PlopTheReal** [comment]
That was a bit tricky I ended up controlling the overflow on drag to disable the scroll (works cross browser and that bloody Safari crap). PreventDefault wasn't working for me for some reasons. Sorry I didn't took time to create a REPL for that as I'm not sure how to make on that will reproduce the issue without spending a lot of time.

**Comment by PlopTheReal** [comment]
Actually you can see the problem (on android) on this REPL; https://svelte.dev/repl/49b4c8b761114d079ce63e51157e1caf?version=3.44.3
If you start dragging it'll also scroll vertically.

**Comment by PlopTheReal** [comment]
Actually it would be nice to have pointerevents within the library as we could then make the distinction between a touch or a pointer down. Then we can use longpress with a 0 timeout when it's a pointer and some delay when it's a touch. Right now I'm trying to use pointerevent to trigger the longpress but I struggle to simulate a mousedown event at the pointer location.

**Comment by PlopTheReal** [comment]
I've updated the following example : https://svelte.dev/repl/49b4c8b761114d079ce63e51157e1caf?version=3.44.3
That's the best solution I've been able to get so far (still playing with the overflow).
The advantages here is that we have less event handlers (one on the root) and we can distinguish a mouse drap and a pointer/touch so when using a mouse you drag straight away and when using a finger you can long press to distinguish both.

Let me know if you think of a better solution like by preventing default wich I couldn't make.

**Comment by isaacHagoel** [comment]
@PlopTheReal Not sure about the problem you are referring to (maybe it doesn't happen to me). Can u attach a short video?

**Comment by PlopTheReal** [comment]
Here is the problem shown on a video: 

https://user-images.githubusercontent.com/58116071/197474862-917dfdc5-9f35-47d7-b7d3-53d22f275137.mp4

Sometimes, the drag triggers also the scroll.

The corresponding REPL is https://svelte.dev/repl/8f71d40b15594bd3b289eb29211062dc?version=3.44.3

**Comment by isaacHagoel** [comment]
for me it doesn't scroll but instead it opens the context menu after a while as if I right clicked.
i think it is because u r dispatching a mousedown event. why?
also, i remember there was a specific reason i didn't use the pointer events api and opted for mousedown and touchstart but i don't remember what it was (could be browser support at the time or maybe something deeper, i vaguely remember that it is much more complex)

**Comment by PlopTheReal** [comment]
I've reduced the time to prevent conflicts with context menu (most of the test are done with chrome deskop latest version but the same is going on with android (chrome also).
I've been dispatching a mousedown because pointerdown will not trigger the drag.
I'll check if that's the same  if I dispatch a touchevent.

I'll also check with touch events (ie no pointer) if the problems stay the same.
The reason I'm using pointer events here is because I can distinguish a touch and a pointer click to reduce the long press to 0 when using a mouse and force a long press when using a finger.

**Comment by PlopTheReal** [comment]
I've updated the REPL with the detection of pointer type: https://svelte.dev/repl/8f71d40b15594bd3b289eb29211062dc?version=3.44.3

Basically thanks to this when using a mouse you can drag straight away and scroll with the wheel.
If you are using touch you can then start dragging with a long press. I think this is quite handy for an app ready for both touches and pointers.

I've checked and the issue stay the same with only touche events of with pointer events and a touch event dispatch to trigger the drag after long press.

**Comment by isaacHagoel** [comment]
a touch event has the equivalent of the mouse event inside of e.touches[0]. this fact can be used to tell them apart.

**Comment by PlopTheReal** [comment]
> a touch event has the equivalent of the mouse event inside of e.touches[0]. this fact can be used to tell them apart.

Nice I didn't knew that. I tends to prefer pointer events as they are supposed to "be the future" from my understanding of W3C things.

On the original issue basically the only way I found to 'fix' in a hacky way is to disable the overflow to prevent any scroll while dragging but that's not ideal. I've tested with preventDefault etc... without success.

**Comment by Supertigerr** [comment]
Why did this issue be closed? Whats the solution to this?

**Comment by heyjoeway** [comment]
I'm not sure if PlopTheReal's playground example worked properly before, but right now it doesn't work on an actual mobile device. It does seem to work in the device preview in Chromium. I really think this issue shouldn't be closed, it makes this entire plugin useless in a lot of cases. If addressed this plugin would be a dream.
